### ISSUE TITLE
router-outlet in loaded component

### ISSUE THREAD

**Issue by fleskberg** [MTobisch/ngx-dynamic-hooks#20]
In my application I´m using router-outlets in multiple levels. In one of the components i want to load with dynamic hooks has router-outlet. The problem is that it takes the first routing component instead of the last, so it kind of tries to start the first component again. Is there a way around this issue or is it not supported?

**Comment by MTobisch** [comment]
I'm not sure I understand fully. There is a router outlet inside one of the dynamically-loaded components and that one is not working correctly or causing the parent outlet to bug out? 

Is there a way you could post a minimal example with the steps to reproduce the bug? Or post the relevant code that is causing this issue?

**Comment by fleskberg** [comment]
Yes, its the router-outlet inside the dynamically loaded component that bugs out, showing incorrect content. It kind of starts over from the beginning of the router chain.

**Comment by MTobisch** [comment]
Okay, this behaviour is new to me. But I have a hard time reproducing it. I've tried a minimal setup: 

- Three top routes
- The second top route has the `ngx-dynamic-hooks` component
- It loads a dynamic component from a string that contains a second, named router-outlet
- Three nested routes for that named outlet

So we have a nested router-outlet inside a component that is dynamically loaded with ngx-dynamic-hooks. But it works fine on my end. I can navigate between the top routes fine and can navigate between the nested routes fine when on top route 2. 

When having (for example) top route 2 + nested route 3 open, and switching top routes, nested route 3 will again render correctly when returning back to top route 2. 

So everything works normally for me. Maybe its something else causing the problem? Have you set up everything right for using named outlets? For example:

- Route config: `{ path: 'someRouteName', component: SomeRouteComponent, outlet: 'customOutlet' }`
- Outlet: `<router-outlet name='customOutlet'></router-outlet>`
- Link: `<a [routerLink]="[{ outlets: { customOutlet: ['someRouteName'] } }]"></a>`

If its not that and you could perhaps post a minimal version of the code that produces the bug, I could also have another look at it.

**Comment by fleskberg** [comment]
Hi! Im not nameing the outlets. But I guess that it would work, so perhaps it’s my workaround.
Before introducing dynamic hooks I used the same component without naming the router outlet without problem.
I can’t send the app I’m working on but perhaps I could make a smaller one with similar logic.

**Comment by fleskberg** [comment]
Made a small example of how i get it to load wrong component, switch in parent component between using dynamic hooks and compare with how it works without.
[dynamic-route.zip](https://github.com/MTobisch/ngx-dynamic-hooks/files/8752043/dynamic-route.zip)

**Comment by MTobisch** [comment]
Hi there, 

thanks for being so kind to provide me with an example project! I encountered the same weird behaviour when running it on my machine.

To cut to the chase, it was indeed a bug with my library which has now been fixed. Here are the patched versions depending on your Angular version:

Angular 6-13: `ngx-dynamic-hooks 1.7.4`
Angular 13+: `ngx-dynamic-hooks 2.0.2`

Since the example your posted ran with Angular 7, you would need to download v1.7.4 of the library. Working with child router-outlets should now work fine.

**The long explanation:** Previously, the library was instantiating dynamic components with the root injector by default for dependency injection. Due to this, when injecting `ActivatedRoute`, it was always pointing to the top route, i.e. the root component of the app, no matter how deeply routed the components actually were.

As `<router-outlet>`s internally consult `ActivatedRoute` when determining which routes to load, the child outlets wrongly assumed they were still at the root level instead of already being a child outlet. Thus they loaded the root routes again. Which load the child outlet again. Which loads the root routes again. And so forth ad infinium.

The patch simply made it so that dynamically-created components now load with the same injector as the `<ngx-dynamic-hooks>`-component (which provides the correct `ActivatedRoute`).
